### PR TITLE
Add a Kotlin Standalone Generator

### PR DESCRIPTION
**Pull request by ghosalmartin**

### CODE REVIEW

File: Sources/KotlinStandAlone/Generator.swift [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+		guard let o = json.swift().data(using: .utf8) else {
+			throw "Failed to generate Swift file"
+		}
+		return o
+	}
+}
+
+private extension Lexicon.Graph.JSON {
+	
+	func swift() -> String {
```
Comment (steprescott) [inline Sources/KotlinStandAlone/Generator.swift:24]:
```suggestion
	func kotlin() -> String {
```

Comment (steprescott) [inline Sources/KotlinStandAlone/Generator.swift]:
```suggestion
	func kotlin() -> String {
```

File: Sources/KotlinStandAlone/Generator.swift [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+import UniformTypeIdentifiers
+
+public enum Generator: CodeGenerator {
+	
+	// TODO: prefixes?
+	
+	public static let utType = UTType.sourceCode
+	
+	public static func generate(_ json: Lexicon.Graph.JSON) throws -> Data {
+		guard let o = json.swift().data(using: .utf8) else {
```
Comment (steprescott) [inline Sources/KotlinStandAlone/Generator.swift]:
```suggestion
		guard let o = json.kotlin().data(using: .utf8) else {
```

File: Sources/KotlinStandAlone/Generator.swift [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+
+"""
+	}
+}
+
+private extension Lexicon.Graph.Node.Class.JSON {
+	
+	// TODO: make this more readable
+	
+	func swift(prefix: (class: String, protocol: String)) -> [String] {
```
Comment (steprescott) [inline Sources/KotlinStandAlone/Generator.swift]:
```suggestion
	func kotlin(prefix: (class: String, protocol: String)) -> [String] {
```

File: Sources/KotlinStandAlone/Generator.swift [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+
+fun <X: I> CallAsFunctionExtension<X>.id(): (I) -> String = { it.identifier }
+
+open class L(override val localized: String = "", override val identifier: String,) : I
+
+// MARK: generated types
+
+val \(name) = L_\(name)("\(name)")
+
+\(classes.flatMap{ $0.swift(prefix: ("L", "I")) }.joined(separator: "\n"))
```
Comment (steprescott) [inline Sources/KotlinStandAlone/Generator.swift]:
```suggestion
\(classes.flatMap{ $0.kotlin(prefix: ("L", "I")) }.joined(separator: "\n"))
```

File: Package.swift [hunk 5]
```diff
@@ -88,5 +89,24 @@ let package = Package(
 				.copy("Resources"),
 			]
 		),
+        
+        // MARK: KotlinStandAlones
+        
+        .target(
+            name: "KotlinStandAlone",
```
Comment (steprescott) [inline Package.swift]: Add this new `KotlinStandAlone` target as a dependency to the `LexiconGenerators` target. Then when you build that target you build all generators.

File: Tests/KotlinStandAloneTests/KotlinStandAlone™.swift [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+		var json = try await "test".taskpaper().lexicon().json()
+		json.date = Date(timeIntervalSinceReferenceDate: 0)
+		
+		let code = try Generator.generate(json).string()
+		        
+		try hope(code) == "test.kt".file().string()
+	}
+	
+	func test_code() throws {
+        // Probably need a Kotlin project/repo for this bit, did test it though and it works fine
```
Comment (steprescott) [inline Tests/KotlinStandAloneTests/KotlinStandAlone™.swift]: LGTM 😂

File: Tests/KotlinStandAloneTests/Resources/test.taskpaper [hunk 7]
```diff
@@ -0,0 +1,17 @@
+test:
```
Comment (steprescott) [inline Tests/KotlinStandAloneTests/Resources/test.taskpaper:1]: Might be good if we can share this file and the util file across test targets. Not one for you to fix in this PR.

**Review by steprescott — CHANGES_REQUESTED** [review]

File: Package.swift [hunk 8]
```diff
@@ -12,6 +12,7 @@ let package = Package(
 		.library(name: "Lexicon", targets: ["Lexicon"]),
 		.library(name: "SwiftLexicon", targets: ["SwiftLexicon"]),
 		.library(name: "SwiftStandAlone", targets: ["SwiftStandAlone"]),
+        .library(name: "KotlinStandAlone", targets: ["KotlinStandAlone"]),
```
Comment (ollieatkinson) [inline Package.swift]: alignment seems wonky here, or is it GitHub?

File: Sources/KotlinStandAlone/Generator.swift [hunk 9]
```diff
@@ -0,0 +1,111 @@
...
+
+public enum Generator: CodeGenerator {
+	
+	// TODO: prefixes?
+	
+	public static let utType = UTType.sourceCode
+	
+	public static func generate(_ json: Lexicon.Graph.JSON) throws -> Data {
+		guard let o = json.swift().data(using: .utf8) else {
+			throw "Failed to generate Swift file"
```
Comment (ollieatkinson) [inline Sources/KotlinStandAlone/Generator.swift]:
```suggestion
			throw "Failed to generate Kotlin file"
```

File: Sources/KotlinStandAlone/Generator.swift (hunk 2 above)
Comment (ollieatkinson) [inline Sources/KotlinStandAlone/Generator.swift]: These days it's better to use the non-failable initialiser for Data:
```suggestion
	let o = Data(json.kotlin().utf8)
```

File: Tests/KotlinStandAloneTests/Resources/test.kt [hunk 10]
```diff
@@ -0,0 +1,61 @@
+interface I: TypeLocalized, SourceCodeIdentifiable
+
+interface TypeLocalized {
+    val localized: String
```
Comment (ollieatkinson) [inline Tests/KotlinStandAloneTests/Resources/test.kt]: How do string localisations work in Android? Are they string-key based?

File: Package.swift (hunk 8 above)
Comment (steprescott) [reply Package.swift]: It's GitHub. When I checked it out it was fine.

**Comment by ollieatkinson** [comment]
It looks like the difference between spaces vs tabs - try enable invisible characters

File: Package.swift (hunk 8 above)
Comment (ghosalmartin) [reply Package.swift]: Same for me unsure wtf was going on

File: Tests/KotlinStandAloneTests/KotlinStandAlone™.swift (hunk 6 above)
Comment (ghosalmartin) [reply Tests/KotlinStandAloneTests/KotlinStandAlone™.swift]: I was thinking I could add the test to the Kotlin generated code

File: Tests/KotlinStandAloneTests/Resources/test.kt (hunk 10 above)
Comment (ghosalmartin) [reply Tests/KotlinStandAloneTests/Resources/test.kt]: Its static xml files. So realistically in our implementation we'd have to sort it ourselves

File: Sources/KotlinStandAlone/Generator.swift (hunk 2 above)
Comment (ghosalmartin) [reply Sources/KotlinStandAlone/Generator.swift]: Done. tbh i just copy pastaed the Swift version so might be worth making the updates there as well

**Comment by ghosalmartin** [comment]
Whats the preferred for Swift? Spaces or tabs?

File: Package.swift (hunk 8 above)
Comment (ghosalmartin) [reply Package.swift]: This is because my entry uses spaces and rest uses tabs

File: Tests/KotlinStandAloneTests/KotlinStandAlone™.swift (hunk 6 above)
Comment (ghosalmartin) [reply Tests/KotlinStandAloneTests/KotlinStandAlone™.swift]: left a comment of how the test should be

**Comment by screensailor** [comment]
For now, I would recommend tabs for this project. Please feel free to reconsider. The reason is that we are likely to continue using string literal small test lexicons and if your IDE is converting tabs to spaces, it's not so much fun... However, we could also develop a dsl so that we don't need string literals. Of course we can (and often should) load test lexicon files, but this is not helping readability of unit tests so much...

File: Sources/KotlinStandAlone/Generator.swift [hunk 11]
```diff
@@ -0,0 +1,101 @@
...
+    }
+    
+}
+
+private extension String {
+    
+    var idToClassSuffix: String {
+        replacingOccurrences(of: "_", with: "__")
+            .replacingOccurrences(of: ".", with: "_")
+            .replacingOccurrences(of: "_&_", with: "_")
```
Comment (DanielNogueiraAndroid) [inline Sources/KotlinStandAlone/Generator.swift]: There is a chance that `replacingOccurrences(of:` "_", with: `"__")` will make all `"_&_" in to "__&__" ` ? If yes is intentional the effect that may have in `.replacingOccurrences(of: "_&_", with: "_")`?  Please ignore if does not make sense.

**Comment by ghosalmartin** [comment]
am just hitting Ctrl+I and thats how it formats it. Can I swap XCode to use tabs?

**Comment by oliveratkinson-bc** [comment]
@ghosalmartin yes, you can make the whole of Xcode use tabs by going to:
> Xcode > Preferences > Text Editing > Indentation

image

or if you want to just do it for this project, you can use the File Inspector on the right  under "Text Settings"

image

**Comment by ghosalmartin** [comment]
All tabbed up, thanks Gents.

**Review by steprescott — APPROVED** [review]

**Pull request merged**